runtime/doc/lua.txt

```diff
@@ -2945,6 +2945,20 @@ Example: >lua
 <
 
 
+vim.fs.abspath({path})                                      *vim.fs.abspath()*
+    Convert path to an absolute path. A tilde (~) character at the beginning
+    of the path is expanded to the user's home directory. Does not check if
```

Comment:
> A tilde (~) character at the beginning of the path is expanded to the user's home directory.

That seems random, and somewhat against the spirit of "Does not check if the path exists".

Expanding the path should perhaps be delegated to [realpath()](https://github.com/neovim/neovim/pull/20976).

Reply:
Though this raises a question about whether `~foo` should also be supported.

Reply:
> That seems random, and somewhat against the spirit of "Does not check if the path exists".

I don't see how it is against it. It doesn't require checking if the path exist, only do some string substitution using OS data. And it does so because it is a common way to denote path starting from user's home directory.

> Though this raises a question about whether ~foo should also be supported.

That arguably should be '~foo' relative path. Perhaps wording in the description could be better with something like "A tilde (~) directory at the beginning of the path ...".

Reply:
> Though this raises a question about whether `~foo` should also be supported.

Could be added later if we want. Both `vim.fs.normalize` and `vim.fs.abspath` use `expand_home` so that function could be modified to add that functionality to both of them

Reply:
> I don't see how it is against it. It doesn't require checking if the path exist, only do some string substitution using OS data.

It implies the existence of a filepath with a home-dir concept, instead of being purely "path only".

Reply:
It does not look like it. It is the same concept as '.' and '..' - convention for a specific directory 

The reasoning of why '\~' is replaced and '.' / '..' are not seems sound: the '\~' denotes "absolute" path while '.' / '..' - relative. Not resolving tilde results in different (pure) path then the input, while not resolving '.' / '..' results in the same path. Having them "squashed" can be done by extra `vim.fs.normalize()` call.

Also, not resolving tilde would require something like `vim.fs.normalize(vim.fs.abspath(vim.fs.normalize(path)))` to ensure that the final result is fully resolved absolute path.

Reply:
> For reference, [python's abspath](https://docs.python.org/3/library/os.path.html#os.path.abspath) doesn't expand "~", it does something braindead:

Python actually does the technically correct thing since you can create a directory named `~`. I guess you can argue if it's more beneficial to denote `~` to mean home directory rather than following all allowable paths.

Reply:
That's a good point. I think that behavior of whether first '~' directory is treated as user's home should then be consistent across all `vim.fs`. For example, `normalize()` resolves it. So ideally they all do or do not resolve by default, while optionally provide a flag to control it. I'd say resolving by default is vastly more useful.

Reply:
I'm unsure at the moment. Optionally supporting expansion for all `vim.fs` functions would mean adding an optional flag for all `vim.fs` function, and I suspect this could massively increase complexity/surface area of the `vim.fs` module. Instead, letting the user resolve the path once via `expand()` or a potential `vim.fs.realpath()` sounds like a simpler method.

Reply:
I don't see why `vim.fs.normalize(vim.fs.abspath(vim.fs.normalize(path)))` is needed.

```lua
local path = "~/programs"
vim.print(vim.fs.normalize(path))
-- /Users/username/programs
```

I might be missing an edge case you found.

Reply:
This isn't about any edge case per se. Assume the plugin gets from the user some `path` and needs to convert it to absolute fully resolved path. As good plugin does, it might want to allow user to at least be able to:
- Use '~/programs' to mean '/home/user/programs'.
- Use '../programs' to mean 'path/to/programs' (with current directory being 'path/to/cwd').

How plugin is expected to achieve that using only `vim.fs` in case `vim.fs.abspath()` doesn't resolve '\~'?
- Using *just* `vim.fs.normalize(path)` doesn't work for '../programs', as it returns '../programs'. But yes, does work for '~/programs' because `vim.fs.normalize()` resolves it (which won't be the case if "... letting the user resolve the path once via expand() or a potential vim.fs.realpath() sounds like a simpler method.")
- Using `vim.fs.abspath(vim.fs.normalize(path))` also doesn't work for '../programs' because it returns '/path/to/cwd/../programs', i.e. not a resolved path.
- Thus it needs to resort to `vim.fs.normalize(vim.fs.abspath(vim.fs.normalize(path)))` or a custom "expand only home directory" function instead of the inner `vim.fs.normalize()`. Which is, of course, pretty straightforward, but a bit annoying.

Reply:
Wouldn't a potential `realpath` function alleviate this problem? Basically a function that says "give me the actual, for real path" where we can blast the path with any amount of path normalization and absolute pathism?

Reply:
> Wouldn't a potential `realpath` function alleviate this problem? Basically a function that says "give me the actual, for real path" where we can blast the path with any amount of path normalization and absolute pathism?

I would assume any function named `realpath()` to work only on "actual, real paths", i.e. paths that are present on disk. This is not always helpful, like for example in cases like "create a file/directory based on the user input `path`" or "act based on user `path` which might or might not actually exist on disk".

Having the ability to work on "pure paths" (i.e. just strings that behave by the rules of file system paths) is pretty liberating as a plugin author. I think it should be at least an active aspiration for `vim.fs` to try and not rely on path actually be present on disk if it is possible to do so.